pallets/teebag/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+			mrenclave: MrEnclave,
+		) -> DispatchResultWithPostInfo {
+			Self::ensure_admin_or_root(origin)?;
+			ensure!(worker_type.is_sidechain(), Error::<T>::ScheduledEnclaveNotApplicable,);
+			ScheduledEnclave::<T>::insert((worker_type, sidechain_block_number), mrenclave);
+			Self::deposit_event(Event::ScheduledEnclaveSet {
+				worker_type,
+				sidechain_block_number,
+				mrenclave,
+			});
+			Ok(().into())
+		}
+
+		#[pallet::call_index(8)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal, Pays::No))]
+		pub fn remove_scheduled_enclave(
+			origin: OriginFor<T>,
+			worker_type: WorkerType,
+			sidechain_block_number: SidechainBlockNumber,
+		) -> DispatchResultWithPostInfo {
+			Self::ensure_admin_or_root(origin)?;
+			ensure!(worker_type.is_sidechain(), Error::<T>::ScheduledEnclaveNotApplicable,);
+			ensure!(
+				ScheduledEnclave::<T>::contains_key((worker_type, sidechain_block_number)),
+				Error::<T>::ScheduledEnclaveNotExist
+			);
+			ScheduledEnclave::<T>::remove((worker_type, sidechain_block_number));
+			Self::deposit_event(Event::ScheduledEnclaveRemoved {
+				worker_type,
+				sidechain_block_number,
+			});
+			Ok(().into())
+		}
+
+		#[pallet::call_index(9)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal, Pays::Yes))]
+		pub fn register_enclave(
+			origin: OriginFor<T>,
+			worker_type: WorkerType,
+			attestation: Vec<u8>,
+			worker_url: Vec<u8>,
+			shielding_pubkey: Option<Vec<u8>>,
+			vc_pubkey: Option<Vec<u8>>,
+			attestation_type: AttestationType,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(worker_url.len() <= MAX_URL_LEN, Error::<T>::EnclaveUrlTooLong);
+
+			let mut enclave = Enclave::new(
+				worker_type,
+				worker_url,
+				shielding_pubkey,
+				vc_pubkey,
+				attestation_type,
+			);
+			match attestation_type {
+				AttestationType::Ignore => {
+					ensure!(
+						Self::mode() == OperationalMode::Development,
+						Error::<T>::InvalidAttestationType
+					);
+					enclave.mrenclave =
+						<MrEnclave>::decode(&mut attestation.as_slice()).unwrap_or_default();
+					enclave.last_seen_timestamp = Self::now().saturated_into();
+					enclave.sgx_build_mode = SgxBuildMode::default();
+				},
+				AttestationType::Ias => {
+					let report = Self::verify_ias(&sender, attestation)?;
+					enclave.mrenclave = report.mr_enclave;
+					enclave.last_seen_timestamp = report.timestamp;
+					enclave.sgx_build_mode = report.build_mode;
+				},
+				AttestationType::Dcap(_) => {
+					let report = Self::verify_dcap(&sender, attestation)?;
+					enclave.mrenclave = report.mr_enclave;
+					enclave.last_seen_timestamp = report.timestamp;
+					enclave.sgx_build_mode = report.build_mode;
+				},
+			};
+
+			match Self::mode() {
+				OperationalMode::Production | OperationalMode::Maintenance => {
+					if !Self::allow_sgx_debug_mode() &&
+						enclave.sgx_build_mode == SgxBuildMode::Debug
+					{
+						return Err(Error::<T>::InvalidSgxMode.into())
```

Comment:
Maybe it should always return error if `OperationalMode::Production && enclave.sgx_build_mode == SgxBuildMode::Debug` ?

Reply:
This is a tricky situation.

This mode exists when you want to test everything including RA report, but using an Intel dev endpoint - it's still a legit RA report, so you could validate all your verification logic. 

We ourselves have such situation on staging where the worker is built with `SGX_PRODUCTION`, parachain without `tee-dev`, so we expect a valid RA report. But, the report comes from a dev endpoint of Intel, which implies `SgxBuildMode == Debug`.

Upstream has this storage `AllowSgxDebugMode` too, I assume they had similar scenario.